In [2]:
import pandas as pd
df = pd.read_csv("mushrooms.csv")
df.head()

,class,cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,...,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat
0,p,x,s,n,t,p,f,c,n,k,...,s,w,w,p,w,o,p,k,s,u
1,e,x,s,y,t,a,f,c,b,k,...,s,w,w,p,w,o,p,n,n,g
2,e,b,s,w,t,l,f,c,b,n,...,s,w,w,p,w,o,p,n,n,m
3,p,x,y,w,t,p,f,c,n,n,...,s,w,w,p,w,o,p,k,s,u
4,e,x,s,g,f,n,f,w,b,k,...,s,w,w,p,w,o,e,n,a,g


In [4]:
df.tail()

,class,cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,...,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat
8119,e,k,s,n,f,n,a,c,b,y,...,s,o,o,p,o,o,p,b,c,l
8120,e,x,s,n,f,n,a,c,b,y,...,s,o,o,p,n,o,p,b,v,l
8121,e,f,s,n,f,n,a,c,b,n,...,s,o,o,p,o,o,p,b,c,l
8122,p,k,y,n,f,y,f,c,n,b,...,k,w,w,p,w,o,e,w,v,l
8123,e,x,s,n,f,n,a,c,b,y,...,s,o,o,p,o,o,p,o,c,l


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

1. Target Variable Analysis (Class)

In [3]:
df['class'].value_counts()


class
e    4208
p    3916
Name: count, dtype: int64

### Target Variable Distribution

The dataset contains two classes:
- Edible (e): 4208 samples
- Poisonous (p): 3916 samples

The edible class is the majority class, while the poisonous class is the minority class.  
However, the difference between the two classes is small, indicating that the dataset is nearly balanced.  
This balance is beneficial for classification tasks as it reduces bias toward a single class.


2.Cap Characteristics Analysis

In [5]:
cap_features = ['cap-shape', 'cap-surface', 'cap-color']

for feature in cap_features:
    print(pd.crosstab(df[feature], df['class']))
    print("\n")


class         e     p
cap-shape            
b           404    48
c             0     4
f          1596  1556
k           228   600
s            32     0
x          1948  1708


class           e     p
cap-surface            
f            1560   760
g               0     4
s            1144  1412
y            1504  1740


class         e     p
cap-color            
b            48   120
c            32    12
e           624   876
g          1032   808
n          1264  1020
p            56    88
r            16     0
u            16     0
w           720   320
y           400   672




Cap features such as cap-shape, cap-surface, and cap-color show noticeable
differences between edible and poisonous mushrooms. Certain cap colors and surface
types appear more frequently in poisonous mushrooms, indicating that cap
characteristics are useful indicators of toxicity.


3. Bruises & Odor Analysis

In [7]:
#(a) Bruises vs Class
pd.crosstab(df['bruises'], df['class'])


class,e,p
bruises,,
f,1456,3292
t,2752,624


The presence or absence of bruises shows a relationship with mushroom edibility.
Mushrooms without bruises are more frequently poisonous, while bruising mushrooms
are more likely to be edible.


In [8]:
#(b) Odor vs Class
pd.crosstab(df['odor'], df['class'])


class,e,p
odor,,
a,400,0
c,0,192
f,0,2160
l,400,0
m,0,36
n,3408,120
p,0,256
s,0,576
y,0,576


Odor is one of the strongest indicators of mushroom toxicity. Certain odors such as
foul, fishy, pungent, and creosote are predominantly associated with poisonous
mushrooms, while edible mushrooms usually have mild or no odor.


4. Gill Features Analysis

In [9]:
gill_features = ['gill-spacing', 'gill-size', 'gill-color']

for feature in gill_features:
    print(pd.crosstab(df[feature], df['class']))
    print("\n")


class            e     p
gill-spacing            
c             3008  3804
w             1200   112


class         e     p
gill-size            
b          3920  1692
n           288  2224


class         e     p
gill-color           
b             0  1728
e            96     0
g           248   504
h           204   528
k           344    64
n           936   112
o            64     0
p           852   640
r             0    24
u           444    48
w           956   246
y            64    22




Gill features including gill-size, gill-spacing, and gill-color vary significantly
between edible and poisonous mushrooms. Certain gill colors and narrow gills are
more common in poisonous mushrooms, making gill characteristics strong predictors
of toxicity.


5. Stalk Characteristics Analysis

In [10]:
stalk_features = [
    'stalk-shape',
    'stalk-root',
    'stalk-surface-above-ring',
    'stalk-surface-below-ring',
    'stalk-color-above-ring',
    'stalk-color-below-ring'
]

for feature in stalk_features:
    print(pd.crosstab(df[feature], df['class']))
    print("\n")


class           e     p
stalk-shape            
e            1616  1900
t            2592  2016


class          e     p
stalk-root            
?            720  1760
b           1920  1856
c            512    44
e            864   256
r            192     0


class                        e     p
stalk-surface-above-ring            
f                          408   144
k                          144  2228
s                         3640  1536
y                           16     8


class                        e     p
stalk-surface-below-ring            
f                          456   144
k                          144  2160
s                         3400  1536
y                          208    76


class                      e     p
stalk-color-above-ring            
b                          0   432
c                          0    36
e                         96     0
g                        576     0
n                         16   432
o                        192     0
p          

Stalk-related features such as stalk-root, stalk-shape, surface texture, and color
show distinct patterns across classes. Certain stalk roots and surface textures
occur more frequently in poisonous mushrooms, indicating their importance in
toxicity identification.


6. Veil & Ring Analysis

In [11]:
veil_ring_features = ['veil-type', 'veil-color', 'ring-number', 'ring-type']

for feature in veil_ring_features:
    print(pd.crosstab(df[feature], df['class']))
    print("\n")


class         e     p
veil-type            
p          4208  3916


class          e     p
veil-color            
n             96     0
o             96     0
w           4016  3908
y              0     8


class           e     p
ring-number            
n               0    36
o            3680  3808
t             528    72


class         e     p
ring-type            
e          1008  1768
f            48     0
l             0  1296
n             0    36
p          3152   816




Veil and ring features show variation between edible and poisonous mushrooms.
Although veil-type is mostly constant, ring-number and ring-type differ across
classes, with some ring patterns appearing predominantly in poisonous mushrooms.


7. Spore, Population & Habitat Analysis

In [12]:
pd.crosstab(df['habitat'], df['class'])


class,e,p
habitat,,
d,1880,1268
g,1408,740
l,240,592
m,256,36
p,136,1008
u,96,272
w,192,0


In [13]:
pd.crosstab(df['population'], df['class'])


class,e,p
population,,
a,384,0
c,288,52
n,400,0
s,880,368
v,1192,2848
y,1064,648


Habitat and population features show meaningful class-wise differences. Certain
habitats have a higher concentration of poisonous mushrooms, indicating that
environmental factors play an important role in mushroom toxicity.


8. Interaction & Combined Effects

In [1]:
#(a) Odor × Gill Color
pd.crosstab([df['odor'], df['gill-color']], df['class'])


NameError: name 'pd' is not defined

In [ ]:
#(b) Cap Color × Habitat
pd.crosstab([df['cap-color'], df['habitat']], df['class'])


Feature combinations such as odor with gill color and cap color with habitat show
stronger relationships with mushroom toxicity than individual features. Certain
combinations are highly indicative of poisonous mushrooms.


9. Feature Engineering Insights

In [ ]:
#(a) Toxic Odor Indicator
toxic_odors = ['f', 'y', 'c', 'p']
df['toxic_odor'] = df['odor'].isin(toxic_odors).astype(int)


In [ ]:
#(b) Cap Feature Combination
df['cap_combo'] = df['cap-shape'] + "_" + df['cap-color']


In [ ]:
#(c) Gill Complexity Score
df['gill_complexity'] = (
    df['gill-size'].map({'b': 1, 'n': 2}) +
    df['gill-spacing'].map({'c': 1, 'w': 2})
)


In [ ]:
#(d) High-Risk Habitat Indicator
high_risk_habitats = ['w', 'u', 'g']
df['high_risk_habitat'] = df['habitat'].isin(high_risk_habitats).astype(int)


New features were engineered to capture hidden patterns in the data, such as toxic
odor indicators, cap feature combinations, gill complexity scores, and high-risk
habitat indicators. These engineered features can improve the performance of
machine learning classification models.


1.Classify features into Qualitative and Quantitative
### Feature Classification: Qualitative vs Quantitative

The mushroom dataset contains only qualitative (categorical) features. All attributes,
including class, cap characteristics, gill features, stalk features, odor, habitat,
and population, represent categories rather than numerical measurements.

There are no quantitative (numerical) features present in the original dataset.
All features are encoded using categorical values.


2. Classify Features into 4 Levels of Measurement
🔹 Nominal Level

Categories with no natural order

Only labels / names

Mushroom features (MOST of them):

class (edible, poisonous)

cap-shape

cap-color

cap-surface

odor

gill-color

habitat

veil-type

veil-color

ring-type

stalk-color-above-ring

stalk-color-below-ring

📌 These features cannot be ordered meaningfully.

🔹 Ordinal Level

Categories with some order

Differences are not measurable

Mushroom features (few):

population (scattered < several < numerous)

gill-size (broad < narrow)

gill-spacing (close < crowded < distant)

ring-number (none < one < two)

📌 Order exists, but no numeric distance.

🔹 Interval Level

Numeric scale

Equal intervals

No true zero

❌ Not present in mushroom dataset

🔹 Ratio Level

Numeric scale

Equal intervals

True zero exists

❌ Not present in mushroom dataset

### Feature Classification Based on Levels of Measurement

The mushroom dataset features can be classified into four measurement levels.
Most features belong to the nominal level as they represent unordered categories
such as color, shape, odor, and habitat. A few features such as population,
gill-size, gill-spacing, and ring-number are ordinal in nature as they have a
natural order.

There are no interval or ratio level features in the mushroom dataset, as it
does not contain any numerical measurements.


3. Descriptive Analysis (Mean, Median, Mode, Quartiles, Std)

⚠️ VERY IMPORTANT CONCEPT

👉 Mean, median, standard deviation, and quartiles are defined ONLY for numerical data.
👉 The mushroom dataset is categorical, so:

❌ Mean → NOT meaningful
❌ Median → NOT meaningful
❌ Q1, Q2, Q3 → NOT meaningful
❌ Standard deviation → NOT meaningful